pandas/io/excel/_base.py

```diff
@@ -860,15 +861,26 @@ class ExcelFile:
     def __init__(
         self, path_or_buffer, engine=None, storage_options: StorageOptions = None
     ):
+        ext = None
+        if not isinstance(path_or_buffer, (BufferedIOBase, RawIOBase)):
+            ext = os.path.splitext(str(path_or_buffer))[-1][1:]
+
         if engine is None:
             engine = "xlrd"
             if isinstance(path_or_buffer, (BufferedIOBase, RawIOBase)):
                 if _is_ods_stream(path_or_buffer):
                     engine = "odf"
             else:
-                ext = os.path.splitext(str(path_or_buffer))[-1]
-                if ext == ".ods":
+                if ext == "ods":
                     engine = "odf"
+
+        elif engine == "xlrd" and ext in ("xlsx", "xlsm"):
```

Comment:
This warning should be in the `if engine is None` branch

Reply:
Are you sure? This will mean that that people also get a warning when they ask for the default (which is still xlrd), instead of when they explicitly ask for xlrd.

Reply:
Yea so the point of it is that people who want to suppress the warnings will get a head start and explicitly request `engine="openpyxl"`, which is a good thing to sniff out any bugs